### PR TITLE
fixed code and README.md file

### PR DESCRIPTION
**Pull request by priyabsah**
## Who worked on the changes?

- 

## New Components/Features

- 

## Enhanced Components/Features

- 

## Other Changes/Fixes

- 

## TODOs

- [ ] Example Checkbox

## Checklist

- [ ] Existing, relevant documentation has been updated to reflect my changes

### CODE REVIEW

File: chatbot_attention_mechanism/chatbot.py [hunk 1]
```diff
@@ -0,0 +1,378 @@
+# -*- coding: utf-8 -*-
+"""chatbot.ipynb
+
+Automatically generated by Colab.
+
+Original file is located at
+    https://colab.research.google.com/drive/1i0Ryq3Wdq-Ej6oMEVMnnJGD-1u3f7Cxb
```
Comment (Manan2711) [inline chatbot_attention_mechanism/chatbot.py]: Can you please remove this url, as it links to your personal colab file.

File: chatbot_attention_mechanism/chatbot.py
Comment (Manan2711) [inline chatbot_attention_mechanism/chatbot.py:1]: Could you please refactor the code and separate the different functionalities into individual files? At the moment, both model training and chatbot execution are in the same file. Moving them into separate files, will make it easier to understand and follow for future teams. Thank you!

**Review by Manan2711 — CHANGES_REQUESTED** [review]
Please make the requested changes. Aside from that, the code looks good. The other error codes can be disregarded as they don't significantly affect the functionality.

**Review by Ashan224019868 — COMMENTED** [review]
LGTM- No Security Issues

File: chatbot_attention_mechanism/chatbot.py
Comment (priyabsah) [reply chatbot_attention_mechanism/chatbot.py:1]: Yes I changed the folder

**Review by Thumbness — DISMISSED** [review]
LGTM, in future, i would also encourage to add line comments explaining "why" this particular function or statement was included, for the benefit of the user.

File: chatbot_attention_mechanism/chatbotpy.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import torch
+#from models.language_model import BigramLanguageModel
+from utils.data_loader import encode
```
Comment (Manan2711) [inline chatbot_attention_mechanism/chatbotpy.py]: Can you please fix this line? The data_loader file does not contain any function named encode.

File: chatbot_attention_mechanism/chatbotpy.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+
+def chatbot_interaction(model, stoi, itos):
```
Comment (Manan2711) [inline chatbot_attention_mechanism/chatbotpy.py:4]: Can you please add a docstring?

File: chatbot_attention_mechanism/training/train_model.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import torch
+
+
+def get_batch(split, data, block_size, batch_size):
```
Comment (Manan2711) [inline chatbot_attention_mechanism/training/train_model.py:3]: Please add a docstring

File: chatbot_attention_mechanism/training/train_model.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    ix = torch.randint(len(data) - block_size, (batch_size,))
+    x = torch.stack([data[i:i + block_size] for i in ix])
+    y = torch.stack([data[i + 1:i + block_size + 1] for i in ix])
+    return x, y
+
+
+@torch.no_grad()
+def estimate_loss(model, train_data, val_data, eval_iters, block_size, batch_size):
```
Comment (Manan2711) [inline chatbot_attention_mechanism/training/train_model.py]: Please add a docstring

File: chatbot_attention_mechanism/training/train_model.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        for k in range(eval_iters):
+            X, Y = get_batch(split, data, block_size, batch_size)
+            _, loss = model(X, Y)
+            losses[k] = loss.item()
+        out[split] = losses.mean()
+    model.train()
+    return out
+
+
+def train_model(model, train_data, val_data, optimizer, max_iters,
```
Comment (Manan2711) [inline chatbot_attention_mechanism/training/train_model.py]: Please add a docstring

File: chatbot_attention_mechanism/training/train_model.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+                eval_interval, block_size, batch_size, eval_iters):
+    for iter in range(max_iters):
```
Comment (Manan2711) [inline chatbot_attention_mechanism/training/train_model.py]: Please rename "iter" as it's a built in python function and shouldn't be redefined.

File: chatbot_attention_mechanism/utils/data_loader.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+import pandas as pd
+
+
+def load_and_preprocess_data():
```
Comment (Manan2711) [inline chatbot_attention_mechanism/utils/data_loader.py:3]: Please add a docstring.

File: chatbot_attention_mechanism/utils/data_loader.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    text1 = text1.drop('Question_ID', axis=1)
+    text2 = text2.drop(['source', 'focus_area'], axis=1)
+    text2 = text2.rename(columns={"question": "Questions", "answer": "Answers"})
+    text = pd.concat([text1, text2], axis=0)
+    text_combined = ''.join(text['Questions'].astype(str) +
+                            text['Answers'].astype(str))
+    return text_combined
+
+
+def encode_text(text):
```
Comment (Manan2711) [inline chatbot_attention_mechanism/utils/data_loader.py:39]: Please add a docstring

**Review by Manan2711 — CHANGES_REQUESTED** [review]
Please go through the added comments and make the changes requested. Thank you!

File: chatbot_attention_mechanism/utils/data_loader.py (hunk 9 above)
Comment (priyabsah) [reply chatbot_attention_mechanism/utils/data_loader.py:39]: Yes, docstring added

File: chatbot_attention_mechanism/utils/data_loader.py (hunk 8 above)
Comment (priyabsah) [reply chatbot_attention_mechanism/utils/data_loader.py:3]: Yes, docstring added.

File: chatbot_attention_mechanism/training/train_model.py (hunk 7 above)
Comment (priyabsah) [reply chatbot_attention_mechanism/training/train_model.py]: Yes, variabe iter is changed.

File: chatbot_attention_mechanism/training/train_model.py (hunk 6 above)
Comment (priyabsah) [reply chatbot_attention_mechanism/training/train_model.py]: Yes, docstrig added.

File: chatbot_attention_mechanism/training/train_model.py (hunk 4 above)
Comment (priyabsah) [reply chatbot_attention_mechanism/training/train_model.py:3]: Yes, docstring added.

File: chatbot_attention_mechanism/chatbotpy.py (hunk 3 above)
Comment (priyabsah) [reply chatbot_attention_mechanism/chatbotpy.py:4]: Yes, added docstring.

File: chatbot_attention_mechanism/main.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    train_model(
+        model=model, 
+        train_data=train_data, 
+        val_data=val_data, 
+        optimizer=optimizer,
+        max_iters=MAX_ITERS, 
+        eval_interval=EVAL_INTERVAL, 
+        block_size=BLOCK_SIZE, 
+        batch_size=BATCH_SIZE, 
+        eval_iters=EVAL_ITERS
```
Comment (Manan2711) [inline chatbot_attention_mechanism/main.py]: The train_model function does not have these 5 parameters. Please fix this. Also the function requires "config" parameter as well.

**Review by Manan2711 — DISMISSED** [review]

File: chatbot_attention_mechanism/main.py (hunk 10 above)
Comment (priyabsah) [reply chatbot_attention_mechanism/main.py]: Yes, fixed this

**Comment by Manan2711** [comment]
The code looks good to be merged! Thank you!

**Review by ben-AI-cybersec — APPROVED** [review]
Lovely torch implementation of an attention mechanism, great work!

**Pull request merged**